sdk/messaging/azwebpubsub/client_custom.go

```diff
@@ -0,0 +1,170 @@
+
+	if err != nil {
+		return nil, err
+	}
+
+	azcoreClient, err := azcore.NewClient(internal.ModuleName+".Client", internal.ModuleVersion, runtime.PipelineOptions{
+		PerRetry: []policy.Policy{internal.NewWebPubSubKeyCredentialPolicy(props.AccessKey)},
+	}, &options.ClientOptions)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		internal: azcoreClient,
+		endpoint: props.Endpoint,
+		hub:      hub,
+		key:      &props.AccessKey,
+	}, nil
+}
+
+// GenerateClientAccessUrlOptions represents the options for generating a client token.
+type GenerateClientAccessUrlOptions struct {
+	// UserID is the user ID for the client.
+	UserID *string
+
+	// Roles are the roles that the connection with the generated token will have.
+	// Roles give the client initial permissions to leave, join, or publish to groups when using PubSub subprotocol.
+	// Possible role values:
+	// - webpubsub.joinLeaveGroup: the client can join or leave any group.
+	// - webpubsub.sendToGroup: the client can send messages to any group.
+	// - webpubsub.joinLeaveGroup.<group>: the client can join or leave group <group>.
+	// - webpubsub.sendToGroup.<group>: the client can send messages to group <group>.
+	// More info: https://azure.github.io/azure-webpubsub/references/pubsub-websocket-subprotocol#permissions
+	Roles []string
+
+	// ExpirationTimeInMinutes is the number of minutes until the token expires.
+	ExpirationTimeInMinutes *int32
+
+	// Groups are the groups to join when the client connects.
+	Groups []string
+}
+
+type GenerateClientAccessUrlResponse struct {
+	// The client token
+	Token string
+	// The base URL for the client to connect to
+	BaseURL string
+	// The URL client connects to with access_token query string
+	URL string
+}
+
+func (c *Client) GenerateClientAccessUrl(ctx context.Context, options *GenerateClientAccessUrlOptions) (*GenerateClientAccessUrlResponse, error) {
+	endpoint := c.endpoint
+	hubName := c.hub
+	if !strings.HasSuffix(endpoint, "/") {
+		endpoint += "/"
+	}
+	clientEndpoint := regexp.MustCompile(`$(http)(s?://)`).ReplaceAllString(endpoint, "ws$2")
+	baseURL := fmt.Sprintf("%sclient/hubs/%s", clientEndpoint, hubName)
+
+	var token string
+	var err error
+	if c.key != nil {
+		token, err = c.signJwtToken(baseURL, options)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// Replace with your logic to generate the token using a webPubSub method
+		resp, err := c.generateClientToken(ctx, hubName, &ClientGenerateClientTokenOptions{UserID: options.UserID, Role: options.Roles, Group: options.Groups, MinutesToExpire: options.ExpirationTimeInMinutes})
+		if err != nil {
+			return nil, err
+		}
+
+		token = *resp.Token
+	}
+
+	return &GenerateClientAccessUrlResponse{
+		Token:   token,
+		BaseURL: baseURL,
+		URL:     fmt.Sprintf("%s?access_token=%s", baseURL, url.QueryEscape(token)),
+	}, nil
+}
+
+func (c *Client) signJwtToken(baseURL string, options *GenerateClientAccessUrlOptions) (string, error) {
+	key := []byte(*c.key)
```

Comment:
Do we need nil check for key? How about use `string` for `c.key` instead of `*string`?

Reply:
when key is nil, we consider this as using AAD token, otherwise we consider this as a connectionString case

Reply:
check nil